File: BotBuilderNodeJs/libraries/botframework-schema/src/sharepoint/getCardViewResponse.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { AceData } from './aceData';
import { ICardParameters } from './ICardParameters';
import { IOnCardSelectionActionParameters } from './IOnCardSelectionActionParameters'
import { ActionButton } from './actionButton';
import { ISharePointViewResponse } from './ISharepointViewResponse';
import { SharepointAction } from '.';

/**
 * Sharepoint GetCardView response object
 */
export class GetCardViewResponse implements ISharePointViewResponse {
    private templateType: GetCardViewResponse.CardViewTemplateType;
    private aceData: AceData;
    private data: ICardParameters;
    private viewId: string;
    private onCardSelection: SharepointAction;
    private cardButtons: ActionButton;

    /**
     * Initializes a new instance of the GetCardViewResponse class
     */
    public GetCardViewResponse(templateType: GetCardViewResponse.CardViewTemplateType) {
        this.TemplateType = templateType;
    }

    /**
     * Sets templateType property of type CardViewTemplateType
     */
    public set TemplateType(templateType: GetCardViewResponse.CardViewTemplateType) {
        this.templateType = templateType;
    }

    /**
     * Gets templateType property of type CardViewTemplateType
     */
    public get TemplateType(): GetCardViewResponse.CardViewTemplateType {
        return this.templateType;
    }

    /**
     * Sets aceData property of type AceData
     */
    public set AceData(aceData: AceData) {
        this.aceData = aceData;
    }

    /**
     * Gets aceData property of type AceData
     */
    public get AceData(): AceData {
        return this.aceData;
    }

    /**
     * Sets data property of type ICardParameters
     */
    public set Data(data: ICardParameters) {
        this.data = data;
    }

    /**
     * Gets data property of type ICardParameters
     */
    public get Data(): ICardParameters {
        return this.data;
    }

    /**
     * Sets on card selection property of type SharepointAction
     */
    public set OnCardSelection(onCardSelection: SharepointAction) {
        this.onCardSelection = onCardSelection;
    }

    /**
     * Gets on card selection property of type SharepointAction
     */
    public get OnCardSelection(): SharepointAction {
        return this.onCardSelection;
    }

    /**
     * Sets card buttons property of type ActionButton
     */
    public set CardButtons(cardButtons: ActionButton) {
        this.cardButtons = cardButtons;
    }

    /**
     * Gets card buttons property of type ActionButton
     */
    public get CardButtons(): ActionButton {
        return this.cardButtons;
    }

    /**
     * Sets viewId property of type CardViewData
     */
    public set ViewId(viewId: string) {
        this.viewId = viewId;
    }

    /**
     * Gets viewId property of type CardViewData
     */
    public get ViewId(): string {
        return this.viewId;
    }
}

export namespace GetCardViewResponse
{
    export enum CardViewTemplateType {
        BasicCardView = "Basic",
        ImageCardView = "Image",
        PrimaryTextCardView = "PrimaryText",
        SignInCardView = "SignIn"
    }
}

File: BotBuilder DotNet/libraries/Microsoft.Bot.Schema/SharePoint/GetCardViewResponse.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Text;
using Microsoft.Bot.Schema.Teams;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Linq;

namespace Microsoft.Bot.Schema.SharePoint
{
    /// <summary>
    /// SharePoint GetCardView response object.
    /// </summary>
    public class GetCardViewResponse : ISharePointViewResponse
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="GetCardViewResponse"/> class.
        /// </summary>
        /// <param name="templateType">Template type of the card view.</param>
        public GetCardViewResponse(CardViewTemplateType templateType)
        {
            this.TemplateType = templateType;
        }

        /// <summary>
        /// This enum contains the different types of card templates available in the SPFx framework.
        /// </summary>
        public enum CardViewTemplateType
        {
            /// <summary>
            /// Primary text card view
            /// </summary>
            [EnumMember(Value = "Basic")]
            BasicCardView,

            /// <summary>
            /// Image card view
            /// </summary>
            [EnumMember(Value = "Image")]
            ImageCardView,

            /// <summary>
            /// Primary Text card view
            /// </summary>
            [EnumMember(Value = "PrimaryText")]
            PrimaryTextCardView,

            /// <summary>
            /// Sign In card view
            /// </summary>
            [EnumMember(Value = "SignIn")]
            SignInCardView
        }

        /// <summary>
        /// Gets or Sets the template type of the card view of type <see cref="CardViewTemplateType"/> enum.
        /// </summary>
        /// <value>This value is the template type of the card view response.</value>
        [JsonProperty(PropertyName = "templateType")]
        [JsonConverter(typeof(StringEnumConverter))]
        public CardViewTemplateType TemplateType { get; set; }

        /// <summary>
        /// Gets or Sets AceData for the card view of type <see cref="AceData"/>.
        /// </summary>
        /// <value>This value is the ace data of the card view response.</value>
        [JsonProperty(PropertyName = "aceData")]
        public AceData AceData { get; set; }

        /// <summary>
        /// Gets or Sets data associated with the card view of type <see cref="ICardParameters"/>.
        /// </summary>
        /// <value>This value is the data of the card view response.</value>
        [JsonProperty(PropertyName = "data")]
        public ICardParameters Data { get; set; }

        /// <summary>
        /// Gets or Sets action to be performed when card is selected of type <see cref="SharepointAction"/>.
        /// </summary>
        /// <value>This value is the action performed when card is clicked.</value>
        [JsonProperty(PropertyName = "onCardSelection")]
        public SharepointAction OnCardSelection { get; set; }

        /// <summary>
        /// Gets or Sets button(s) on the card view of type <see cref="ActionButton"/>.
        /// </summary>
        /// <value>This value is the button(s) associated with the card view.</value>
        [JsonProperty(PropertyName = "cardButtons")]
        public IEnumerable<ActionButton> CardButtons { get; set; }

        /// <summary>
        /// Gets or Sets the view Id of type <see cref="string"/>.
        /// </summary>
        /// <value>This value is the view id of the card view.</value>
        [JsonProperty(PropertyName = "viewId")]
        public string ViewId { get; set; }
    }
}
